{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "kernelspec": {
      "display_name": "R",
      "language": "R",
      "name": "ir"
    },
    "language_info": {
      "codemirror_mode": "r",
      "file_extension": ".r",
      "mimetype": "text/x-r-source",
      "name": "R",
      "pygments_lexer": "r",
      "version": "3.3.1"
    },
    "colab": {
      "name": "PRSenrichQTL_tutorial.ipynb",
      "provenance": [],
      "collapsed_sections": []
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "7oOMohGzCXft",
        "colab_type": "text"
      },
      "source": [
        "# PRS-preselection-based trans QTL analysis\n",
        "\n",
        "# Summary\n",
        "\n",
        "This is a short summary of our polygenic risk score (PRS) based pathway enrichment approach to preselect candidate genes for trans QTL analyses.\n",
        "\n",
        "You can find a preprint version of the manuscript  \n",
        "Ines Assum & Julia Krause et al., Tissue-specific multiOMICs analysis of atrial fibrillation, *bioRxiv* (2020)  \n",
        "here:\n",
        "https://doi.org/10.1101/2020.04.06.021527\n",
        "\n",
        "We address a key hypothesis about the existence of core genes as postulated in the omnigenic model by [Liu et al.](https://doi.org/10.1016/j.cell.2019.04.014), *Cell* (2019). Core genes are central genes with trans-associations to GWAS loci, whose expression levels directly affect a disease phenotype. Here we sought to identify candidate core genes for AF to understand the contribution of trans-genetic effects in the pathology of AF. To prioritize genes satisfying the properties predicted by the omnigenic model, we evaluated the accumulation of trans-effects, their relevance in gene regulatory networks, and the disease association by the following strategy:\n",
        "\n",
        "\n",
        "i) Evaluate cumulated trans-effects of disease-associated variants on expression by ranking genes based on their correlation of mRNA and protein abundance with a PRS (eQTS) as proposed by [Võsa et al.](https://doi.org/10.1101/447367), *bioRxiv* (2018)\n",
        "\n",
        "ii) Identify genes sharing molecular function and representing biological networks that propagate genetic trans-effects to core genes by pathway enrichment analysis (GSEA) on the eQTS rankings. Genes driving the enrichment of multiple gene sets were selected as core gene candidates.\n",
        "\n",
        "iii) Establish the link between the core gene candidates and the disease based on a significant trans eQTL GWAS hit.\n",
        "\n",
        "# Example\n",
        "\n",
        "## Requirements\n",
        "\n",
        "This tutorial uses the following packages:\n",
        "\n",
        "* [fgsea](https://bioconductor.org/packages/release/bioc/html/fgsea.html)\n",
        "* [MatrixEQTL](https://github.com/andreyshabalin/MatrixEQTL)\n",
        "* [eQTLpipeline](https://github.com/matthiasheinig/eQTLpipeline)\n",
        "\n",
        "and data\n",
        "\n",
        "* [1k_genomes_PRS_expr.tsv](https://github.com/heiniglab/symatrial/blob/master/example_data/1k_genomes_PRS_expr.tsv)\n",
        "* [1k_genomes_AF_snps.txt](https://github.com/heiniglab/symatrial/blob/master/example_data/1k_genomes_AF_snps.txt)\n",
        "* [GO bp .gmt file](https://www.gsea-msigdb.org/gsea/msigdb/download_file.jsp?filePath=/msigdb/release/7.1/c5.bp.v7.1.symbols.gmt).\n",
        "\n",
        "Packages are installed in the first code chunk (running it might take up to 15 minutes) and the sample dataset is downloaded in the second.\n",
        "\n",
        "You only need to download gene set annotations for the GO biological processes ([GO bp .gmt file](https://www.gsea-msigdb.org/gsea/msigdb/download_file.jsp?filePath=/msigdb/release/7.1/c5.bp.v7.1.symbols.gmt)).\n",
        "To upload the .gmt file, click on the folder symbol on the top left. Using the panel that opens, you can choose \"Upload\" to open a window to browse your local file system."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "0dXVDkyWCXfy",
        "colab_type": "code",
        "outputId": "a9984225-fb2e-4be5-a5ec-b9ed18b9bf17",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 972
        }
      },
      "source": [
        "#Install necessary libraries\n",
        "install.packages(c(\"devtools\",\"RColorBrewer\"), dependencies = T, clean = T)\n",
        "BiocManager::install(\"fgsea\", dependencies = T, clean = T)\n",
        "\n",
        "library(devtools)\n",
        "devtools::install_github(\"andreyshabalin/MatrixEQTL\", force=T)\n",
        "devtools::install_github(\"matthiasheinig/eQTLpipeline\", force=T)"
      ],
      "execution_count": 1,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Installing packages into ‘/usr/local/lib/R/site-library’\n",
            "(as ‘lib’ is unspecified)\n",
            "\n",
            "also installing the dependencies ‘clisymbols’, ‘gargle’, ‘cyclocomp’, ‘xmlparsedata’, ‘highlight’, ‘parsedate’, ‘rappdirs’, ‘whoami’, ‘hunspell’, ‘BiocManager’, ‘curl’, ‘foghorn’, ‘gmailr’, ‘lintr’, ‘mockery’, ‘pingr’, ‘pkgdown’, ‘rhub’, ‘spelling’\n",
            "\n",
            "\n",
            "Bioconductor version 3.10 (BiocManager 1.30.10), R 3.6.3 (2020-02-29)\n",
            "\n",
            "Installing package(s) 'BiocVersion', 'fgsea'\n",
            "\n",
            "also installing the dependencies ‘formatR’, ‘bit’, ‘lambda.r’, ‘futile.options’, ‘bit64’, ‘blob’, ‘futile.logger’, ‘snow’, ‘BiocGenerics’, ‘Biobase’, ‘IRanges’, ‘RSQLite’, ‘S4Vectors’, ‘data.table’, ‘BiocParallel’, ‘gridExtra’, ‘fastmatch’, ‘reactome.db’, ‘AnnotationDbi’, ‘org.Mm.eg.db’, ‘limma’, ‘GEOquery’\n",
            "\n",
            "\n",
            "Old packages: 'backports', 'curl', 'ellipsis', 'farver', 'glue', 'modelr',\n",
            "  'processx', 'rex', 'rprojroot', 'systemfonts', 'tidyr', 'tinytex', 'xfun'\n",
            "\n",
            "Loading required package: usethis\n",
            "\n",
            "Downloading GitHub repo andreyshabalin/MatrixEQTL@master\n",
            "\n"
          ],
          "name": "stderr"
        },
        {
          "output_type": "stream",
          "text": [
            "\u001b[32m✔\u001b[39m  \u001b[90mchecking for file ‘/tmp/Rtmpf1ao1k/remotes787b596895/andreyshabalin-MatrixEQTL-681b367/DESCRIPTION’\u001b[39m\u001b[36m\u001b[39m\n",
            "\u001b[90m─\u001b[39m\u001b[90m  \u001b[39m\u001b[90mpreparing ‘MatrixEQTL’:\u001b[39m\u001b[36m\u001b[39m\n",
            "\u001b[32m✔\u001b[39m  \u001b[90mchecking DESCRIPTION meta-information\u001b[39m\u001b[36m\u001b[39m\n",
            "\u001b[90m─\u001b[39m\u001b[90m  \u001b[39m\u001b[90mchecking for LF line-endings in source and make files and shell scripts\u001b[39m\u001b[36m\u001b[39m\n",
            "\u001b[90m─\u001b[39m\u001b[90m  \u001b[39m\u001b[90mchecking for empty or unneeded directories\u001b[39m\u001b[36m\u001b[39m\n",
            "\u001b[90m─\u001b[39m\u001b[90m  \u001b[39m\u001b[90mlooking to see if a ‘data/datalist’ file should be added\u001b[39m\u001b[36m\u001b[39m\n",
            "\u001b[90m─\u001b[39m\u001b[90m  \u001b[39m\u001b[90mbuilding ‘MatrixEQTL_2.3.tar.gz’\u001b[39m\u001b[36m\u001b[39m\n",
            "   \n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "stream",
          "text": [
            "Installing package into ‘/usr/local/lib/R/site-library’\n",
            "(as ‘lib’ is unspecified)\n",
            "\n",
            "Downloading GitHub repo matthiasheinig/eQTLpipeline@master\n",
            "\n"
          ],
          "name": "stderr"
        },
        {
          "output_type": "stream",
          "text": [
            "BatchJobs (NA    -> 1.8  ) [CRAN]\n",
            "checkmate (NA    -> 2.0.0) [CRAN]\n",
            "backports (1.1.6 -> 1.1.7) [CRAN]\n",
            "ellipsis  (0.3.0 -> 0.3.1) [CRAN]\n",
            "glue      (1.4.0 -> 1.4.1) [CRAN]\n",
            "BBmisc    (NA    -> 1.11 ) [CRAN]\n",
            "sendmailR (NA    -> 1.2-1) [CRAN]\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "stream",
          "text": [
            "Installing 7 packages: BatchJobs, checkmate, backports, ellipsis, glue, BBmisc, sendmailR\n",
            "\n",
            "Installing packages into ‘/usr/local/lib/R/site-library’\n",
            "(as ‘lib’ is unspecified)\n",
            "\n"
          ],
          "name": "stderr"
        },
        {
          "output_type": "stream",
          "text": [
            "\u001b[32m✔\u001b[39m  \u001b[90mchecking for file ‘/tmp/Rtmpf1ao1k/remotes784174166b/matthiasheinig-eQTLpipeline-c4f2f5b/DESCRIPTION’\u001b[39m\u001b[36m\u001b[39m\n",
            "\u001b[90m─\u001b[39m\u001b[90m  \u001b[39m\u001b[90mpreparing ‘eQTLpipeline’:\u001b[39m\u001b[36m\u001b[39m\n",
            "\u001b[32m✔\u001b[39m  \u001b[90mchecking DESCRIPTION meta-information\u001b[39m\u001b[36m\u001b[39m\n",
            "\u001b[90m─\u001b[39m\u001b[90m  \u001b[39m\u001b[90mchecking for LF line-endings in source and make files and shell scripts\u001b[39m\u001b[36m\u001b[39m\n",
            "\u001b[90m─\u001b[39m\u001b[90m  \u001b[39m\u001b[90mchecking for empty or unneeded directories\u001b[39m\u001b[36m\u001b[39m\n",
            "\u001b[90m─\u001b[39m\u001b[90m  \u001b[39m\u001b[90mbuilding ‘eQTLpipeline_0.1.tar.gz’\u001b[39m\u001b[36m\u001b[39m\n",
            "   \n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "stream",
          "text": [
            "Installing package into ‘/usr/local/lib/R/site-library’\n",
            "(as ‘lib’ is unspecified)\n",
            "\n"
          ],
          "name": "stderr"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "Q0dOdnpZ1Z9Z",
        "colab_type": "code",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 51
        },
        "outputId": "595434e3-d342-4887-8509-ef93d7ccdc41"
      },
      "source": [
        "#Download data set\n",
        "download.file(\"https://raw.githubusercontent.com/heiniglab/symatrial/master/example_data/1k_genomes_PRS_expr.tsv\",\n",
        "              destfile = \"1k_genomes_PRS_expr.tsv\")\n",
        "download.file(\"https://raw.githubusercontent.com/heiniglab/symatrial/master/example_data/1k_genomes_AF_snps.txt\",\n",
        "              destfile = \"1k_genomes_AF_snps.txt\")\n",
        "dir.create(\"/tmp/\")"
      ],
      "execution_count": 4,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Warning message in dir.create(\"/tmp/\"):\n",
            "“'/tmp' already exists”\n"
          ],
          "name": "stderr"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "uuScCL3dFbFQ",
        "colab_type": "code",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "outputId": "fca8f2bf-a1e9-4386-d588-11bb425203c5"
      },
      "source": [
        "#Load libraries\n",
        "library(fgsea)\n",
        "library(MatrixEQTL)\n",
        "library(eQTLpipeline)\n",
        "\n",
        "#R session info including all loaded libraries\n",
        "sessionInfo()"
      ],
      "execution_count": 6,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Loading required package: Rcpp\n",
            "\n",
            "Loading required package: data.table\n",
            "\n",
            "Loading required package: BatchJobs\n",
            "\n",
            "Loading required package: BBmisc\n",
            "\n",
            "\n",
            "Attaching package: ‘BBmisc’\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:base’:\n",
            "\n",
            "    isFALSE\n",
            "\n",
            "\n",
            "The development of BatchJobs and BatchExperiments is discontinued.\n",
            "\n",
            "Consider switching to 'batchtools' for new features and improved stability\n",
            "\n",
            "Sourced 1 configuration files: \n",
            "\n",
            "  1: /usr/local/lib/R/site-library/BatchJobs/etc/BatchJobs_global_config.R\n",
            "\n",
            "BatchJobs configuration:\n",
            "  cluster functions: Interactive\n",
            "  mail.from: \n",
            "  mail.to: \n",
            "  mail.start: none\n",
            "  mail.done: none\n",
            "  mail.error: none\n",
            "  default.resources: \n",
            "  debug: FALSE\n",
            "  raise.warnings: FALSE\n",
            "  staged.queries: TRUE\n",
            "  max.concurrent.jobs: Inf\n",
            "  fs.timeout: NA\n",
            "  measure.mem: TRUE\n",
            "\n",
            "\n"
          ],
          "name": "stderr"
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "R version 3.6.3 (2020-02-29)\n",
              "Platform: x86_64-pc-linux-gnu (64-bit)\n",
              "Running under: Ubuntu 18.04.3 LTS\n",
              "\n",
              "Matrix products: default\n",
              "BLAS:   /usr/lib/x86_64-linux-gnu/openblas/libblas.so.3\n",
              "LAPACK: /usr/lib/x86_64-linux-gnu/libopenblasp-r0.2.20.so\n",
              "\n",
              "locale:\n",
              " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
              " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
              " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
              " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
              " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
              "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
              "\n",
              "attached base packages:\n",
              "[1] stats     graphics  grDevices utils     datasets  methods   base     \n",
              "\n",
              "other attached packages:\n",
              "[1] eQTLpipeline_0.1  BatchJobs_1.8     BBmisc_1.11       data.table_1.12.8\n",
              "[5] MatrixEQTL_2.3    fgsea_1.12.0      Rcpp_1.0.4.6      devtools_2.3.0   \n",
              "[9] usethis_1.6.1    \n",
              "\n",
              "loaded via a namespace (and not attached):\n",
              " [1] lattice_0.20-41     prettyunits_1.1.1   ps_1.3.3           \n",
              " [4] assertthat_0.2.1    rprojroot_1.2       digest_0.6.25      \n",
              " [7] IRdisplay_0.7.0     R6_2.4.1            repr_1.1.0         \n",
              "[10] backports_1.1.6     RSQLite_2.2.0       evaluate_0.14      \n",
              "[13] ggplot2_3.3.0       pillar_1.4.4        rlang_0.4.6        \n",
              "[16] curl_4.3            uuid_0.1-4          blob_1.2.1         \n",
              "[19] callr_3.4.3         Matrix_1.2-18       checkmate_2.0.0    \n",
              "[22] desc_1.2.0          BiocParallel_1.20.1 bit_1.1-15.2       \n",
              "[25] munsell_0.5.0       sendmailR_1.2-1     compiler_3.6.3     \n",
              "[28] pkgconfig_2.0.3     base64enc_0.1-3     pkgbuild_1.0.8     \n",
              "[31] htmltools_0.4.0     tidyselect_1.1.0    tibble_3.0.1       \n",
              "[34] gridExtra_2.3       fansi_0.4.1         crayon_1.3.4       \n",
              "[37] dplyr_0.8.5         withr_2.2.0         grid_3.6.3         \n",
              "[40] jsonlite_1.6.1      gtable_0.3.0        lifecycle_0.2.0    \n",
              "[43] DBI_1.1.0           magrittr_1.5        scales_1.1.1       \n",
              "[46] stringi_1.4.6       cli_2.0.2           fs_1.4.1           \n",
              "[49] remotes_2.1.1       testthat_2.3.2      brew_1.0-6         \n",
              "[52] ellipsis_0.3.0      vctrs_0.3.0         fastmatch_1.1-0    \n",
              "[55] IRkernel_1.1        tools_3.6.3         bit64_0.9-7        \n",
              "[58] glue_1.4.0          purrr_0.3.4         processx_3.4.1     \n",
              "[61] pkgload_1.0.2       parallel_3.6.3      colorspace_1.4-1   \n",
              "[64] BiocManager_1.30.10 sessioninfo_1.1.1   memoise_1.1.0      \n",
              "[67] pbdZMQ_0.3-3       "
            ]
          },
          "metadata": {
            "tags": []
          }
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "NdURZ8SMIPDQ",
        "colab_type": "text"
      },
      "source": [
        "## Loading data\n",
        "\n",
        "In the following, we show the general procedure based on public data (272 individuals from the 1000 genomes LCL data).  \n",
        "We supply a small test dataset with pre-computed PRS values for AF and CAD as well as gene expression for 830 genes (`1k_genomes_PRS_expr.tsv`). Additionally, genotypes for 109 pruned AF SNPs for those individuals are supplied (`1k_genomes_AF_snps.txt`) in MatrixEQTL format."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "27GnAJsuIWaz",
        "colab_type": "code",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 221
        },
        "outputId": "80e01089-7f47-45fd-8b55-b811f8b7fd50"
      },
      "source": [
        "df <- read.csv(file = \"1k_genomes_PRS_expr.tsv\",\n",
        "               sep = \"\\t\", h = T, stringsAsFactors = F)\n",
        "print.data.frame(df[1:5, 1:10],\n",
        "                 row.names = F, digits = 3)"
      ],
      "execution_count": 5,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "      id PRS.AF PRS.AF.percentile PRS.CAD PRS.CAD.percentile sex population\n",
            " HG00096   32.5             0.876    18.0              0.104   1        GBR\n",
            " HG00097   32.5             0.837    18.3              0.986   2        GBR\n",
            " HG00099   32.4             0.563    18.2              0.790   2        GBR\n",
            " HG00100   32.4             0.604    18.0              0.173   2        GBR\n",
            " HG00101   32.3             0.331    18.1              0.544   1        GBR\n",
            "  BAD DBNDD1 CD38\n",
            " 3.74   4.57 4.22\n",
            " 3.50   3.33 4.16\n",
            " 2.48   4.60 3.99\n",
            " 1.95   3.22 4.39\n",
            " 3.50   3.26 4.88\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "LWxZ80Ey_-yH",
        "colab_type": "text"
      },
      "source": [
        "## eQTS ranking\n",
        "First step is to calculate the eQTS rankings:"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "Qc83wlqZAAew",
        "colab_type": "code",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 272
        },
        "outputId": "eda3f457-4999-461d-91b2-bb0520b5cfee"
      },
      "source": [
        "eQTS <- data.frame(matrix(nrow = 830, ncol = 5),\n",
        "                   stringsAsFactors = F,\n",
        "                   row.names = colnames(df)[-c(1:7)])\n",
        "colnames(eQTS) <- c(\"id\", \"Estimate\", \"StdError\", \"tvalue\", \"pvalueT\")\n",
        "data <- df[, c(\"id\", \"PRS.AF\", \"PRS.AF.percentile\",\n",
        "               \"PRS.CAD\", \"PRS.CAD.percentile\",\n",
        "               \"sex\", \"population\")]\n",
        "for (i in 8:(dim(df)[2])){\n",
        "    #i=8\n",
        "    id <- colnames(df)[i]\n",
        "    data$expr <- df[, id]\n",
        "    lmres <- lm(expr ~ sex + PRS.AF.percentile,\n",
        "                data = data)\n",
        "    eQTS[id, ] <- c(id,\n",
        "                    summary(lmres)$coefficients[\"PRS.AF.percentile\", ])\n",
        "    if(i %% 100 == 0) print(paste0(\"Gene \", i-7, \" / \", dim(df)[2]-7, \" done\"))\n",
        "}\n",
        "eQTS[, -1] <- apply(eQTS[, -1], 2, as.numeric)\n",
        "eQTS <- eQTS[order(eQTS$pvalueT), ]\n",
        "\n",
        "print(\"Top five associated genes:\")\n",
        "print.data.frame(eQTS[1:5, ],\n",
        "                 row.names = F, digits = 3)"
      ],
      "execution_count": 7,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "[1] \"Gene 93 / 830 done\"\n",
            "[1] \"Gene 193 / 830 done\"\n",
            "[1] \"Gene 293 / 830 done\"\n",
            "[1] \"Gene 393 / 830 done\"\n",
            "[1] \"Gene 493 / 830 done\"\n",
            "[1] \"Gene 593 / 830 done\"\n",
            "[1] \"Gene 693 / 830 done\"\n",
            "[1] \"Gene 793 / 830 done\"\n",
            "[1] \"Top five associated genes:\"\n",
            "           id Estimate StdError tvalue  pvalueT\n",
            " NAALADL2.AS2    1.019   0.1696   6.01 6.03e-09\n",
            "      BCL2L10    0.458   0.0968   4.73 3.64e-06\n",
            "         CLMN    0.465   0.1052   4.42 1.42e-05\n",
            "   AL606807.1    0.474   0.1080   4.39 1.65e-05\n",
            "       KIF21A    0.383   0.0933   4.11 5.31e-05\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "kX-g6x-9ALKO",
        "colab_type": "text"
      },
      "source": [
        "## Load pathway annotations\n",
        "\n",
        "As mentioned before, please download GO biological processes gene sets first, [GO bp .gmt file](https://www.gsea-msigdb.org/gsea/msigdb/download_file.jsp?filePath=/msigdb/release/7.1/c5.bp.v7.1.symbols.gmt),\n",
        "then import them into R using the fgsea package."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "H7ihUaWxAW7M",
        "colab_type": "code",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 34
        },
        "outputId": "fa6ef8ee-a3f7-4777-e067-5ad9fbf82eee"
      },
      "source": [
        "if(!file.exists(\"c5.bp.v7.1.symbols.gmt\")){\n",
        "  cat(\"Please download pathway annotations first to move on! \n",
        "  To upload the .gmt file, click on the folder symbol on the top left. \n",
        "  Using the panel that opens, you can choose 'Upload' to open a window to \n",
        "  browse your local file system.\")\n",
        "} else {\n",
        "  gmt <- gmtPathways(\"c5.bp.v7.1.symbols.gmt\")\n",
        "  print(\"GO biological processes gene set annotations loaded.\")\n",
        "}"
      ],
      "execution_count": 8,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "[1] \"GO biological processes gene set annotations loaded.\"\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "4GbDL4q_BCyD",
        "colab_type": "text"
      },
      "source": [
        "## Compute gene set enrichment analysis on eQTS ranking:\n"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "KwyD8yfSBHok",
        "colab_type": "code",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 200
        },
        "outputId": "1ba5addd-6a90-4928-c80e-f81224c44419"
      },
      "source": [
        "fgsea.res <- NULL\n",
        "try({\n",
        "  set.seed(1234)\n",
        "  rank <- eQTS[, \"tvalue\"]\n",
        "  names(rank) <- eQTS[, \"id\"]\n",
        "  fgsea.res <- fgsea(gmt,\n",
        "                     rank,\n",
        "                     nperm=10000,\n",
        "                     minSize = 5,\n",
        "                     maxSize=100)\n",
        "  fgsea.res <- fgsea.res[order(fgsea.res$pval), ]\n",
        "})\n",
        "  \n",
        "if(is.null(fgsea.res)){\n",
        "  print(\"fgsea is not installed or pathway annotations are missing. Let's load the results:\")\n",
        "  fgsea.res <- readRDS(\"https://github.com/heiniglab/symatrial/raw/master/example_data/1k_genomes_fgsea_eQTS.RDS\")\n",
        "}\n",
        "\n",
        "print(\"Top five pathways:\")\n",
        "head(fgsea.res[1:5, -8])"
      ],
      "execution_count": 9,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "[1] \"Top five pathways:\"\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "  pathway                                                     pval        \n",
              "1 GO_REGULATION_OF_MEGAKARYOCYTE_DIFFERENTIATION              0.0009973404\n",
              "2 GO_MEGAKARYOCYTE_DIFFERENTIATION                            0.0009973404\n",
              "3 GO_CYTOKINE_PRODUCTION                                      0.0012408231\n",
              "4 GO_POSITIVE_REGULATION_OF_CYTOKINE_PRODUCTION               0.0012794541\n",
              "5 GO_NEGATIVE_REGULATION_OF_INTRACELLULAR_SIGNAL_TRANSDUCTION 0.0046130781\n",
              "  padj      ES         NES       nMoreExtreme size\n",
              "1 0.4884316 -0.9102749 -2.069582  2            5  \n",
              "2 0.4884316 -0.9102749 -2.069582  2            5  \n",
              "3 0.4884316  0.4738125  1.607792 11           75  \n",
              "4 0.4884316  0.5201614  1.680610 11           49  \n",
              "5 0.5895384  0.5952123  1.687400 39           22  "
            ],
            "text/latex": "A data.table: 5 × 7\n\\begin{tabular}{lllllll}\n pathway & pval & padj & ES & NES & nMoreExtreme & size\\\\\n <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <int>\\\\\n\\hline\n\t GO\\_REGULATION\\_OF\\_MEGAKARYOCYTE\\_DIFFERENTIATION              & 0.0009973404 & 0.4884316 & -0.9102749 & -2.069582 &  2 &  5\\\\\n\t GO\\_MEGAKARYOCYTE\\_DIFFERENTIATION                            & 0.0009973404 & 0.4884316 & -0.9102749 & -2.069582 &  2 &  5\\\\\n\t GO\\_CYTOKINE\\_PRODUCTION                                      & 0.0012408231 & 0.4884316 &  0.4738125 &  1.607792 & 11 & 75\\\\\n\t GO\\_POSITIVE\\_REGULATION\\_OF\\_CYTOKINE\\_PRODUCTION               & 0.0012794541 & 0.4884316 &  0.5201614 &  1.680610 & 11 & 49\\\\\n\t GO\\_NEGATIVE\\_REGULATION\\_OF\\_INTRACELLULAR\\_SIGNAL\\_TRANSDUCTION & 0.0046130781 & 0.5895384 &  0.5952123 &  1.687400 & 39 & 22\\\\\n\\end{tabular}\n",
            "text/markdown": "\nA data.table: 5 × 7\n\n| pathway &lt;chr&gt; | pval &lt;dbl&gt; | padj &lt;dbl&gt; | ES &lt;dbl&gt; | NES &lt;dbl&gt; | nMoreExtreme &lt;dbl&gt; | size &lt;int&gt; |\n|---|---|---|---|---|---|---|\n| GO_REGULATION_OF_MEGAKARYOCYTE_DIFFERENTIATION              | 0.0009973404 | 0.4884316 | -0.9102749 | -2.069582 |  2 |  5 |\n| GO_MEGAKARYOCYTE_DIFFERENTIATION                            | 0.0009973404 | 0.4884316 | -0.9102749 | -2.069582 |  2 |  5 |\n| GO_CYTOKINE_PRODUCTION                                      | 0.0012408231 | 0.4884316 |  0.4738125 |  1.607792 | 11 | 75 |\n| GO_POSITIVE_REGULATION_OF_CYTOKINE_PRODUCTION               | 0.0012794541 | 0.4884316 |  0.5201614 |  1.680610 | 11 | 49 |\n| GO_NEGATIVE_REGULATION_OF_INTRACELLULAR_SIGNAL_TRANSDUCTION | 0.0046130781 | 0.5895384 |  0.5952123 |  1.687400 | 39 | 22 |\n\n",
            "text/html": [
              "<table>\n",
              "<caption>A data.table: 5 × 7</caption>\n",
              "<thead>\n",
              "\t<tr><th scope=col>pathway</th><th scope=col>pval</th><th scope=col>padj</th><th scope=col>ES</th><th scope=col>NES</th><th scope=col>nMoreExtreme</th><th scope=col>size</th></tr>\n",
              "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
              "</thead>\n",
              "<tbody>\n",
              "\t<tr><td>GO_REGULATION_OF_MEGAKARYOCYTE_DIFFERENTIATION             </td><td>0.0009973404</td><td>0.4884316</td><td>-0.9102749</td><td>-2.069582</td><td> 2</td><td> 5</td></tr>\n",
              "\t<tr><td>GO_MEGAKARYOCYTE_DIFFERENTIATION                           </td><td>0.0009973404</td><td>0.4884316</td><td>-0.9102749</td><td>-2.069582</td><td> 2</td><td> 5</td></tr>\n",
              "\t<tr><td>GO_CYTOKINE_PRODUCTION                                     </td><td>0.0012408231</td><td>0.4884316</td><td> 0.4738125</td><td> 1.607792</td><td>11</td><td>75</td></tr>\n",
              "\t<tr><td>GO_POSITIVE_REGULATION_OF_CYTOKINE_PRODUCTION              </td><td>0.0012794541</td><td>0.4884316</td><td> 0.5201614</td><td> 1.680610</td><td>11</td><td>49</td></tr>\n",
              "\t<tr><td>GO_NEGATIVE_REGULATION_OF_INTRACELLULAR_SIGNAL_TRANSDUCTION</td><td>0.0046130781</td><td>0.5895384</td><td> 0.5952123</td><td> 1.687400</td><td>39</td><td>22</td></tr>\n",
              "</tbody>\n",
              "</table>\n"
            ]
          },
          "metadata": {
            "tags": []
          }
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Qxybeb5bBa1p",
        "colab_type": "text"
      },
      "source": [
        "## Extract the leadingEdge as candidate genes:\n",
        "\n",
        "As this is only an example dataset for demonstration, we choose a significane threshold of P < 0.01. However, for a real analysis we would recommend a much more stringent cutoff based on the adjusted p-value (FDR).\n",
        "\n",
        "For each gene set, the leadingEdge contains the genes that possibly drive the enrichment. We therefore use those genes for significant pathways. To narrow it down further, we check if the same genes are contained in the leadingEdge of multiple pathways.  \n",
        "We take all genes, that appear more than 5 times (8 genes in total) as candidates for the trans eQTL analysis."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "nPDDasH0BbR4",
        "colab_type": "code",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 471
        },
        "outputId": "8babfcca-5ac2-495d-c930-a84236ae2840"
      },
      "source": [
        "if(file.exists(\"c5.bp.v7.1.symbols.gmt\")){\n",
        "  lead <- data.frame(table(unlist( fgsea.res[fgsea.res$pval<0.01, \"leadingEdge\"])))\n",
        "  hist(lead$Freq, breaks = 20,\n",
        "       main = \"Distribution: Frequencies of genes in multiple leading edges:\")\n",
        "  print(\"Let's select the top candidates that appear more than 5 times:\")\n",
        "  candidates <- as.character(lead[lead$Freq>5, \"Var1\"])\n",
        "  print(candidates)\n",
        "}"
      ],
      "execution_count": 10,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "[1] \"Let's select the top candidates that appear more than 5 times:\"\n",
            "[1] \"CD226\"  \"CD274\"  \"CD28\"   \"IGHD\"   \"IL10\"   \"IL23R\"  \"RGCC\"   \"TNFSF4\"\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "display_data",
          "data": {
            "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAMAAADKOT/pAAADAFBMVEUAAAABAQECAgIDAwME\nBAQFBQUGBgYHBwcICAgJCQkKCgoLCwsMDAwNDQ0ODg4PDw8QEBARERESEhITExMUFBQVFRUW\nFhYXFxcYGBgZGRkaGhobGxscHBwdHR0eHh4fHx8gICAhISEiIiIjIyMkJCQlJSUmJiYnJyco\nKCgpKSkqKiorKyssLCwtLS0uLi4vLy8wMDAxMTEyMjIzMzM0NDQ1NTU2NjY3Nzc4ODg5OTk6\nOjo7Ozs8PDw9PT0+Pj4/Pz9AQEBBQUFCQkJDQ0NERERFRUVGRkZHR0dISEhJSUlKSkpLS0tM\nTExNTU1OTk5PT09QUFBRUVFSUlJTU1NUVFRVVVVWVlZXV1dYWFhZWVlaWlpbW1tcXFxdXV1e\nXl5fX19gYGBhYWFiYmJjY2NkZGRlZWVmZmZnZ2doaGhpaWlqampra2tsbGxtbW1ubm5vb29w\ncHBxcXFycnJzc3N0dHR1dXV2dnZ3d3d4eHh5eXl6enp7e3t8fHx9fX1+fn5/f3+AgICBgYGC\ngoKDg4OEhISFhYWGhoaHh4eIiIiJiYmKioqLi4uMjIyNjY2Ojo6Pj4+QkJCRkZGSkpKTk5OU\nlJSVlZWWlpaXl5eYmJiZmZmampqbm5ucnJydnZ2enp6fn5+goKChoaGioqKjo6OkpKSlpaWm\npqanp6eoqKipqamqqqqrq6usrKytra2urq6vr6+wsLCxsbGysrKzs7O0tLS1tbW2tra3t7e4\nuLi5ubm6urq7u7u8vLy9vb2+vr6/v7/AwMDBwcHCwsLDw8PExMTFxcXGxsbHx8fIyMjJycnK\nysrLy8vMzMzNzc3Ozs7Pz8/Q0NDR0dHS0tLT09PU1NTV1dXW1tbX19fY2NjZ2dna2trb29vc\n3Nzd3d3e3t7f39/g4ODh4eHi4uLj4+Pk5OTl5eXm5ubn5+fo6Ojp6enq6urr6+vs7Ozt7e3u\n7u7v7+/w8PDx8fHy8vLz8/P09PT19fX29vb39/f4+Pj5+fn6+vr7+/v8/Pz9/f3+/v7////i\nsF19AAAACXBIWXMAABJ0AAASdAHeZh94AAAgAElEQVR4nO3dCXwU9dnA8X/uA2KMAiJHAJVq\ntS0RtNgqlgrFAxBaWy1UC4IVxQN90SJeWH2FVl+1Fq13teKJiq/VeiFCxVusbVVsjehboAio\nIIKAIWTemdnd7MxkZrK7eZKZnfl9Px+zm+yzO/+dzY89siZKA9BmKugFAFFASIAAQgIEEBIg\ngJAAAYQECCAkQAAhAQIICRBASIAAQgIEEBIggJAAAYQECCAkQAAhAQIICRBASIAAQgIEEBIg\ngJAAAYQECCAkQAAhAQIICRBASIAAQgIEEBIggJAAAYQECCAkQAAhAQIICRBASIAAQgIEEBIg\ngJAAAYQECCAkQEBHh/SQUmWZDmU0m+FUcHJe3/yB5RV7bRVejaHNe8xxAe6X57OVjBfwqFJF\n2cy3cjntqZ1Duk0ZCnbd76QHG8wvhCOkxLISBmd75izl+l3wqrm6zdLL0QipXXRISKba540v\n1F933dwWU2uL1HvWz5NDrey/5LlcLzHzZbV7SLmszzBFqerr7m6QXo6W+4qapW4Y31tAMKQ2\nLzgaIR0zZvShnfTDwke9pn6n7CEltbK/Pc6V+bISfpnjZbS3o5WaEvQaPKRuGN9bQDCkNotG\nSBv1w+1zK5SqXO0xdWhOIXmcK5tlhdkwpaYHvQYPqRvG9xYgJEnp79inC5Q6Pb0Pd9wyrGtx\n14Nmf6JpIxMPsaZpdyh1+I6zu3SzPEeq0BYM2aVq6GLjLBclH4Yl9krzuZpvlU1XDq4p6Tbi\nD43GJ/plDdFeGL5rp8MWmie+/NBDL7stS0sOJzasaf+csk9Z1UHX7zC/Pu/gyt1GvvEPfXib\nc/P2UefWtNXnfr2yfP8Z6zXLd431oi1XP816BaYkH3imnyP9YZCxmA8Ti/HfeitbSq6oxdlS\n+2KI9lBdRZ+LGrTlo3ft9IN3Wuz65AU4b4Hbldpfe/DQXaqGv2TZimOXap4nNd0/omtx1bd/\nZ9582p2DKmuOeeN/bc+RWizYsUu0ltfefjl+O9G+o+zfLpnosJC0Y5Xq1pTaJw1Dk98oe32Q\nvkHuU+pb1xhXOR1S55vM0wqf0FoN6W89kxf5nU/1zx5Q6hvPlBqfFj1nnDpGqTGuyzI0b1h7\npDxxGUdsS2xQVzZb/9Bi8/ZR59aer06ctsc/0t811nnr1W9muwItQjo3sZgbE4vx3XprW0qu\nyHm2BOOr8wuMr57xYRfjoOvnGYZ0j1I9rja/VPyM5n697Quwn/Sz5EJHNemfTE9c24ttITkX\n7NglSbYLtV+O30607yj7t0smOi6ku/Sjy1P7RO9jv/tffvo4pb6nvfuYfsq9Sz80Tuvbu6Ru\nX+urdlWn3H+V/l3Z+yvnrWk9lzG8Qf827HfT/84oVmpk4rx79q2bOUIf+rZxJt+Qmjf8of4A\n9Px/vfE9pWZq2lv6t9OBt937vdLEreD4ZrKNOrb28W769Xr4njql9t2RWp9t3nr1U+xX4IOl\nByn106VLdyZPfV2/6Lrb5n23c+tbb3VL6d1r20XNJ+7Z69gz9X1ednSPaYP1E6/zCMl5Czyo\nH5RNuH+2vsQ+DZrr9bYvwHbS4/o/l79/+w792s/XtNeM7+4Ffzqq2BaSY8GOXZJku1DH5fjt\nRPuOCnNIr+hHn0/tk4lKXaMfNIw789c7tY9V4rG2fprqv0qz7jh1mn74sH74ZItb03IuY/hX\nSu3yH/1wnv7lZYnzDtlm/jtXaLzy1VpIiQ2fqdRQ/eCTzqpqmzZVqZrPNG1bP9eQbKOOrc1U\nqsuXmrZev6EeTq3PNm+7+kmOK+B4jnSqUrvqd1Rb+7S+9Va3ZNm91l2U3hc/0f+x1g/KV2hf\n7avUaI+QnLeAcc5JmnnHoZ7RXK+3fQG2k24YOXKaZj5u+bmm/UK/I9yiP9razxmSdcGOXZJk\nu1DH5fjtRPuOCnNIb+tHH0vtk7OVqr17bXLIGtJ9mmbbcW/rh427KHVpayENUGqicWpjjVKX\nJ85rvN7+nH74f67LSr1qd69lw3srddE23eFKPacdYN6mmjbLNSTbqGNr31DqFGP0paeeqk+t\nzzZvu/pJjivgCGl/pU4yDi9pfeutbsmye1vuIuOrr+vfUGVKjdM/PU//NzyLkP6mH+6oUuoy\nzfV62xfgdtJZSo3QtK8nktSuaBGSZcGOXZJku1DH5fjtRLcdlY2OC+kF/egLqX3yVqXxcHTv\nyY8YTy2tIZlXJb3jSsx/Rgea32O+ITUVJ/5F0bTvmN8BxmV9oX9Wn4yx5bJSLkpvuKkw/eXr\nNX2Js43hh91Cso/at9ZUpNSc5m0l12ebt139BOcVcIRkX4zv1lvbkvX7suUuMr5qvJ2iV+I6\n6E8d9848pMTNdaBSJ7tfb59doi0cs1eZeXyYplUkd+CCFiFZFuy4fZK70Xahjsvx24luOyob\nHRfS78x/R1LPMxcfkLgGfV+23SBF5i2R3nG7mmc9TKkTWglpi/7ZLebwcP35avr57KoMQzI3\nvMXy5VlNqZv+abeQbKOOrRmn3dy8rfT60vO2q5/gvAL2kByLaW3rvluyPgVtuYtSX9X/ub5J\nP5ibVUjWm8v1envvkt/rHzt9fUAXIyTnrnddsHMovRu9bkLfnei2o7LRcSF9V6m9LK98Nr18\n+VHGa1v6Y1jHDWLbccZLOMY/cZPMW7PO+OwWr3ukq80z68+Pf5ZJSPbnSOawcVfy2+avlif/\nOZvfHJJl8/ZR+9Z2FqbuXNKn2edtVz/5BccVcNwjlSUX81CrW291S7mFZN31niElbr7EzeV+\nvb12yWb9DmG8fk841bxHKk/edcz3CcmxS1LX1nkTWi/Hbye67ahsdFhI9ynziU7z3jY0PlZj\nvIzgE5L5deNBt/6sYY75+rn5yrDLc6S6xLef1qAP/ybHkLSvGa/mpuifTDAOL0xsz7F526hj\na/smnyPde8UVT6VOs83brn6S4wo4Qton+fD+4ta33uqWsg/Jcd29nyPV64c7OqeepXpcb7eT\njAf+byWu9TDzFPPp6Uy/kBy7JMl5E1ovx28nuu2obHRQSI03leqpb0jtjK2zJx5rPogbodSj\n2lp9aKnmHtK5mvm6qvEk03g563FNe68ysRsc57pcf2RgvPB2q1IF/2r5XfLqo4++2nJZSc0b\n1p/q9vhSX+34ky9YrU1QqkZf8eZeie05Nm8bdWztXKW6fK5pG/Qb5bbUadZ5+9VPclwBR0gn\nKlX9ib6Y3q1vvdUtZR+S47onZxy3gHFzXaAlXvB7XnO73o5tWE9aqMzXON7V78wP14xdv6tl\n17sv2LFLkpw3ofVyfHaiY0fZv10y0SHvtRszzPjRXskizXr/cdyTy174VYkqW6c1lig1ZP4z\nLULSAyoqPfexa3ZRal/9KeCKAv1bbcp5u303sRsc59qo76x9bnzk/LLEvYHzu6S1l78TG16h\nPzs99M/P/EipAxq1JfrQoAfmDU7+1MGxeduoY2v/1u9UBj947yClarekTrPN265+kuMKOEJa\npF/yt+664+BOrW+91S1lH5LjuidnHLeAflBcMmvx73dP3Fxu19u+DetJ/9ELGv32n3rqd+a7\nvLIusev/eLC+6ws9F+zYJUkuN2Hz5fjtRPuOCuXL30m9jH+8Ujvj7V7JrxbeoZnvzzR+CukM\n6R79AcWV5lS5+aaT08zj/V/U/8luanmu5jcG/GiblmtI2kOJF45UT+Mhy8nm0crfJG8ox+at\no86tPVGZOK3739KnWeftVz/JfgWc77WbYF+M39Zb21L2ITmue2rGfgvoBzUXpG8ut+vt2Ib1\npLPMYz0+6qGMp/6JXd/pBv3DTs8FO3ZJ6sJb3oSpy/HbifYdFdaQSvY8+qbEj+NSO2PtFQft\nUVK536l/Nz5ZPXbX8n5XtgjpdqX20+6sK6859m/mlxvn9C/teeq61foFfulyrk3/fXB1yZ4/\n/JNtO1mGpC2f1K+s8hsXbTCO77xm37I9jn/7qeRud2zeOtrie/LD0/tXVBww8xPrhVsv2nb1\nU2xXwBnSzqv3Lev243886XJpLbbeypZyCMl+3VMz9ltAP+ik3WLcXH+37VTrahz723JSw2/2\nr+h5yn+0hfsW93rA2PWl3X78zrvKfMHbY8HOXeJ+E1oux28n2ndU6EKKgqfs/+IF64/6v9pB\nr8HDQ45v6I6S/S5pj51ISK0KRUjL50z9qfEm5WOVGhv0Wjx0dEjZ75L23ImE1KpQhFSvP90f\nu+TFc/THKM8GvRYPHR1S9rukPXciIbUqFCFplyWfC6tLgl6Jlw5/aJf9LmnHnUhIrQpHSNqi\nH/cqKetzwpKg1+Gp458jZb9L2m8nEhIggJAAAYQECCAkQAAhAQIICRBASIAAQgIEEBIggJAA\nAYQECCAkQAAhAQIICRBASIAAQgIEEBIggJAAAYQECCAkQAAhAQIICRBASIAAQgIEEBIggJAA\nAYQECCAkQAAhAQIICRBASIAAQgIEEBIggJAAAYQECCAkQAAhAQIICRBASIAAQgIEEBIggJAA\nAYQECCAkQAAhAQIICRBASIAAQgIEEBIggJAAAYQECCAkQAAhAQIICRBASIAAQgIEEBIgIMiQ\n1izM0JoAFwlkIsiQJpfWZKR0coCLBDIRZEgTJ8rOAYEhJEAAIQECCAkQQEiAAEICBBASIICQ\nAAGEBAggJEAAIQECCAkQQEiAAEICBBASIICQAAGEBAggJEAAIQECCAkQQEiAAEICBBASIICQ\nAAH5ENL40csys6F91wt4yoeQvqYy9Iv2XS/gKR9C2mcf2csDxBESIICQAAGEBAggJEAAIQEC\nCAkQQEiAAEICBBASIICQAAGEBAggJEAAIQEC2hJS04qFCxYsWpnz+QkJkZF7SBumd0v8X0C1\nl2/N7RIICZGRc0hr+qn+E2ddddXF43qoAbn9r6mEhMjIOaTJJfOTxxpvLJiW00UQEiIj55C6\nT0ofP6F3ThdBSIiMnEMquTJ9/LLSnC6CkBAZOYfU5/j08TF9c7oIQkJk5BzStIKrtyeObblU\nzcjpIggJkZFzSBsHqqphE888Y8LQSjVkc04XQUiIjNx/jvTVtXVFxo+RSg65tTG3SyAkREab\n3iK07f0336z/yu2Ud62/ANXr7ISEyBB5r92Gjxxf+KDA+gtQGzzORkiIjNxD+vsxfQ67MfGg\nbkaLS/liQ7Onlet9lkZIiJCcQ3qxTFWWqO+Zbw5qGZLFS4SE6Ms5pJEljzZtv7bk4C0aIQE5\nh9T7ROPjotJjGgkJyP0tQpeaB3erswkJyDmkXscmDmeqqwgJsZdzSGcXzDVf1m6aoM45i5AQ\nczmH9GmtGm4eaTpbKUJCzOX+c6RPpp6TPPbI3oSEmGv/3yJESIgBQgIEEBIggJAAAYQECCAk\nQAAhAQIICRBASIAAQgIEEBIggJAAAYQECCAkQAAhAQIICRBASIAAQgIEEBIggJAAAYQECCAk\nQAAhAQIICRBASIAAQgIEEBIggJAAAYQECCAkQAAhAQIICRBASIAAQgIEEBIggJAAAYQECCAk\nQAAhAQIICRBASIAAQgIEEBIggJAAAYQECCAkQAAhAQIICRBASIAAQgIEEBIggJAAAYQECCAk\nQAAhAQIICRBASIAAQgIEEBIggJAAAYQECCAkQAAhAQIICRBASIAAQgIEEBIggJAAAYQECCAk\nQAAhAQIICRBASIAAQgIEEBIggJAAAYQECCAkQAAhAQIICRBASIAAQgIEEBIggJAAAYQECCAk\nQAAhAQIICRBASIAAQgIEEBIggJAAAYQECCAkQAAhAQIICRBASIAAQgIEEBIggJAAAYQECCAk\nQAAhAQIICRBASIAAQgIEEBIggJAAAYQECCAkQEBbQ/rq9ec/9J8gJMRAziFd8bzx8eYapdSg\nt/wGCQkxkHNIaob+4QlV9sMph6rqD3wGCQkx0LaQ+lcv1z8+UnCyzyAhIQbaFNJ6daF5fGxP\nn0FCQgy0KaSVap55/OISn0FCQgy0KaTG6jnm8Um7+QwSEmIg95DGvVH/ycx9vtSPvtdptM8g\nISEGcg8p4WFNu7dT4es+g4SEGMg5pDuvmzVtwtihizTtxp6P+w0SEmJA4C1Cm3e2+NLK/fZq\n1kNt9zgjISEyRN5r92m94wsNf7yl2S+5R0L0iYQ0w+9SeGiHGCAkQAAhAQJyDmmQRXdCQszl\nHFJhYVmzIkJCzOUc0oyq9Et1PLRD3OUcUsOBBzWkjhMS4i73FxuWV5yXOkpIiLs2vGq36bPU\nsSVzfMYICTHAbxECBBASIICQAAGEBAggJEAAIQECCAkQQEiAAEICBBASIICQAAGEBAggJEAA\nIQECCAkQQEiAAEICBBASIICQAAGEBAggJEAAIQECCAkQQEiAAEICBBASIICQAAGEBAggJEAA\nIQECCAkQQEiAAEICBBASIICQAAGEBAggJEAAIQECCAkQQEiAAEICBBASIICQAAGEBAggJEAA\nIQECCAkQYA3pkJs/b4ctEBJiwBpSsaoY9+xO6S0QEmLAGtKntwwrUr0vqpfdAiEhBhzPkdbf\n9P1CddjtXwhugZAQAy1fbFhz3QBVedq/xLZASIiBFiFtfei4ClVbUnJZk9AWCAkx4AjpxVN2\nURU/W6ytPE7NEtoCISEGrCGt/O/+Sh14w0bjeNPwbkJbICTEgDWkQlV92rLUJzcUCG2BkBAD\n1pCG3LU1/Un9AqEtEBJiwP4c6Z1PjA9/Fd0CISEGrCE1TFKL9YO5amKj4BYICTFgDekaNfJD\n/eCfJ6jfCm6BkBAD1pC+OSp55JgMv3MzQkiIAWtIFdckj1xVIrgFQkIMWEPa46zkkal7CG6B\nkBAD1pAmVf7ZOGi4tfgkwS0QEmLAGtKaPVXtD0Ydtpva89+CWyAkxIDt50hrT9tdKdX1F6sl\nt0BIiAHHm1ab/vPBFuEtEBJigF9+AgiwhtQ0f1TdAQmCWyAkxIA1pKuVqqxOENwCISEGrCH1\nOnJFO2yBkBAD1pBKXm2PLRASYsB2j/RKe2yBkBAD1pDOn9oeWyAkxIA1pM1Hjn96eb1JcAuE\nhBiwhqTSBLdASIgBazLjJkxOEdwCISEGeGcDIMAR0hfvbJTeAiEhBmwhLRmk1FOaNvo5yS0Q\nEmLAGtJrpVVH6iGt7166zHM+e4SEGLCGNLJ21cfGPdK62jGCWyAkxIA1pN3naGZI2uwawS0Q\nEmLA9qcv70mGdCe/RQjIiu29dhclQzq5j+AWCAkxYA3p1Jo3jZA2XKgk33RHSIgBa0gf9y4e\nqOrqylTtWsEtEBJiwPZzpHWnG79FqMvp6yS3QEiIAedvEVpbL3lvZCAkxADvtQMEWEMa1myI\n4BYICTHg+v8jVfUQ3AIhIQasIe0wffnOeYdvEtwCISEGXJ8jXXCa4BbCF9ITp2boicwuD3AP\n6ZVoP7Sb+LXMOvoa93DIlGtIz1YKbiGEIQnPAbaQNiasX1wX7d/9TUgQ5/5bhOYJboGQEAO2\n/7EvYezpEf9fzQkJ4uL4zgZCgjhCavscYAtpwLcHWwltgZAQA9aQ9qhQShXo/1UUGYS2QEiI\nAWtIGw4746/btE1/+dGIaL9FiJAgzhrSyalvnKNOEdwCISEGrCF1vSN55H+6CW6BkBAD1pDK\nrkwe+WWZ4BYICTFgDenAHok/IvtilwGCWyAkxIA1pMeKVL/ho4fvpQoeFtwCISEG7H+N4shy\npVTpEQszO2/TioULFixa2coUISEGHO9s2Ln6/VWNmZ1zw/RuiXe41l6+1W+OkBADjpAy/0Nj\na/qp/hNnXXXVxeN6qAEbfAYJCTFgf2iXxR8am1wyP3ms8caCaT6DhIQYsIaU1R8a6z4pffyE\n3j6DhIQYsP3/SNn8obGSK9PHLyv1GSQkxIA1pKz+0Fif49PHx/T1GSQkxIA1pKz+0Ni0gqu3\nJ45tuVTN8BkkJMSANaSs/tDYxoGqatjEM8+YMLRSDdnsM0hIiAFrSNn9obGvrq0rMn6MVHLI\nrb4/eSIkxIA1pKz/0Ni29998s94tk00Xzmh2IiEh+mw/R8r1D419Wu/4wvpxP2l2BCEh+hzv\nbMjtD43N8PsVKjy0QwzY3v39To4XQkiIO2sC5b/O8UIICXFnTWD40TszP+Mgi+6EhJizJrB2\n3FH3Las3tX7GwsKyZkWEhJizJpD+Jfp+YSTNqErXxkM7xJ01gRNOmjQ5qfUzNhx4UEPqOCEh\n7jK47/GwvOK81FFCQtw1JzB3qXnw1uqMz7rps9SxJXN8xggJMdAckkr8X67qDOktEBJigJDa\nPgcQksAcQEgCcwAhCcwBhCQwBxCSwBxgCWnwLIM62DwQ3AIhIQbSIdkIboGQEAPNycyzEdwC\nISEGJO973BESYoCQ2j4HEJLAHEBIAnMAIQnMAYQkMAcQksAcQEgCcwAhCcwBhCQwBxCSwBxA\nSAJzACEJzAGEJDAHEJLAHEBIAnMAIQnMAYQkMAcQksAcQEgCcwAhCcwBhCQwBxCSwBxASAJz\nACEJzAGEJDAHEJLAHEBIAnMAIQnMAYQkMAcQksAcQEgCcwAhCcwBhCQwBxCSwBxASAJzACEJ\nzAGEJDAHEJLAHEBIAnMAIQnMAYQkMAcQksAcQEgCcwAhCcwBhCQwBxCSwBxASAJzACEJzAGE\nJDAHEJLAHEBIAnMAIQnMAYQkMAcQksAcQEgCcwAhCcwBhCQwBxCSwBxASAJzACEJzAGEJDAH\nEJLAHEBIAnMAIQnMAYQkMAcQksAcQEgCcwAhCcwBhCQwBxCSwBxASAJzACEJzAGEJDAHEJLA\nHEBIAnMAIQnMAYQkMAcQksAcQEgCcwAhCcwBhCQwBxCSwBxASAJzACEJzAGEJDAHEJLAHEBI\nAnMAIQnMAYQkMAcQksAcQEgCcwAhCcwBhCQwBxCSwBxASAJzACEJzAGEJDAHEJLAHEBIAnMA\nIQnMAYQkMAcQksAcQEgCcwAhCcwBbQ7pq9ef/9B/gpAQAzmHdMXzxseba5RSg97yGyQkxEDO\nIakZ+ocnVNkPpxyqqj/wGSQkxEDbQupfvVz/+EjByT6DhIQYaFNI69WF5vGxPR0nNv1lYbPf\nRj6kdQsztC6zy0MealNIK9U88/jFJY4TV5Qpi+0eFxGVkE5RGTols8tDHmpTSI3Vc8zjk3bz\nGYz+QzseKqINIY17o/6Tmft8qR99r9Non0FCynYOeSj3kBIe1rR7OxW+7jNISNnOIQ/lHNKd\n182aNmHs0EWadmPPx/0GCSnbOeQhgbcIbd7pezIhZTuHPMR77cI3hzxESOGbQx4ipPDNIQ8R\nUvjmkIcIKXxzyEOEFL455CFCCt8c8hAhhW8OeYiQwjeHPERI4ZtDHiKk8M0hDxFS+OaQhwgp\nfHPIQ4QUvjnkIUIK3xzyECGFbw55iJDCN4c8REjhm0MeIqTwzSEPEVL45pCHCCl8c8hDhBS+\nOeQhQgrfHPIQIYVvDnmIkMI3hzxESOGbQx4ipPDNIQ8RUvjmkIcIKXxzyEOEFL455CFCCt8c\n8hAhhW8OeYiQwjeHPERI4ZtDHiKk8M0hDxFS+OaQhwgpfHPIQ4QUvjnkIUIK3xzyECGFbw55\niJDCN4c8REjhm0MeIqTwzSEPEVL45pCHCCl8c8hDhBS+uW0rMrQtmMuDC0IK39xUlaGpwVwe\nXBBSCOeOy+wO5LiALg8uCIm5bOfggpCYy3YOLgiJuWzn4IKQmMt2Di4Iibls5+CCkJjLdg4u\nCIm5bOfggpCYy3YOLgiJuWzn4IKQmMt2Di4Iibls5+CCkJjLdg4uCIm5bOfggpCYy3YOLgiJ\nuWzn4IKQmMt2Di4Iibls5+CCkJjLdg4uCIm5bOfggpCYy3YOLgiJuWzn4IKQmMt2Di4Iibls\n5+CCkJjLdq5xQ4YaM7s8aYGsj5CYy3ZuWqa/uXVaZpcnLZD1ERJzWc+NWpaRUQE9VAxkfYTE\nXNBz0gJZHyExF/ScNELyQEjRnpNGSB4IKdpz0gjJAyFFe04aIXkgpGjPSSMkD4QU7TlphOSB\nkKI9J42QPBBStOekEZIHQor2nDRC8kBI0Z6TRkgeCCnac9IIyQMhRXtOGiF5IKRoz0kjJA+E\nFO05aYTkgZCiPSeNkDwQUrTnpBGSB0KK9pw0QvJASNGek0ZIHggp2nPSCMkDIUV7ThoheSCk\naM9JIyQPhBTtOWmE5IGQoj0njZA8EFK0587P9Dejnh/M+jJCSMwFPif8m1EJyQMhMRfkXEYI\niTnmBBASc8wJICTmmBNASMwxJ4CQmGNOACExx5wAQmKOOQGExBxzAgiJOeYEEBJzzAkgJOaY\nE0BIzDEngJCYY04AITHHnABCYo45AYTEHHMCCIk55gQQEnPMCWhLSE0rFi5YsGhlK1OExFx+\nz2Uk95A2TO+W+N0utZdv9ZsjJObyey4jOYe0pp/qP3HWVVddPK6HGrDBZ5CQmMvvuYzkHNLk\nkvnJY403FkzzGSQk5vJ7LiM5h9R9Uvr4Cb19BgmJufyey0jOIZVcmT5+WanjxA+71jSrUg0e\nFzG5tCYjhYWZzRWVZTZXVsTlcXk1NaWTc/3md5FzSH2OTx8f09dx4s7FC5s9e4/XRaxZmJkH\nH2SOOfm5hWty/eZ3kXNI0wqu3p44tuVSNUNqOUB+yjmkjQNV1bCJZ54xYWilGrJZcklA/sn9\n50hfXVtXZPwYqeSQWxsFFwTkoza9RWjb+2++We/1mhwQI+3/XjsgBggJEEBIgABCAgQQEiCA\nkAABhAQIICRAACEBAggJEEBIgABCAgQQEiCAkAABhAQIICRAACEBAqIU0iEKkXZI0N9hPqIU\n0vjRy0JtNOtrk9Hjg/4O8xGlkER/c2Y7YH1tE+r1EVLHYX1tE+r1EVLHYX1tE+r1EVLHYX1t\nE+r1EVLHYX1tE+r1EVLHYX1tE+r1EVLHYX1tE+r1EVLHYX1tE+r1EVLHYX1tE+r1EVLHYX1t\nE+r1RSmkU08NegX+WF/bhCpUW0MAAAaLSURBVHp9UQppw4agV+CP9bVNqNcXpZCAwBASIICQ\nAAGEBAggJEAAIQECCAkQQEiAAEICBBASIICQAAGEBAggJEAAIQECCAkQQEiAgOiEtGF6bWnf\nMa8EvQxf56rJQS/B25OHd67+/uKgV+HpvRO7F3cZ+1rQy/ASmZA+66tGXvKz4vJ/BL0QH28U\nhTikP6i9Lz6va+lLQa/DwztVu1169xXdixcFvRAPkQnpDDVX//iIOibohXjbUTcgvCGt63zg\nFk2r7zw16IV4GK+e1z/+XQ0NeiEeIhPSOcMa9I9NFX2CXoi3Xxc8Fd6QrlZPGwdNQa/Dy2Bl\n3L7aLn2DXoiHyISUsL3k0KCX4OmDitM3hjekIysatO2bgl6Ftwnqbf3jJ4VHB70QDxEL6Xrz\nAV44Ddvz8xCH1Gf/vx5aoPa+M+h1eFleM2Dpx38dVvlq0AvxEK2QlpQetiPoNXi5Uz2shTik\nqj57Tn/4+lp1b9AL8fLP/ZVStS8HvQwvkQrpvrKBnwW9Bi/rdhulhTmkMvVH/eOazt0bg16J\nu+X9el/z+B0HVC8MeiEeIhRS06XqqC+CXoSnn3b+d6hD2r3oS+PgJyqkPz84pHK1/vHLnj0b\ngl6Ju+iE1DRJnRXSf011T6pLVq1a9a4atyqkT+gHFZnfoVNVOH+QtLng++bhz9U7Aa/EQ3RC\nmqZmB70EH9NVyoygl+LuTGU+jR+hVga9Elfr1XfMw+PVsoBX4iEyIT2ipgW9BD/LHzc8oEY8\n/l7QS3G3rOCI7Zr2RuG3gl6Ih34l/9I/btxtl+1Br8RdZELaW501wxTm37Qe4udI2jmq7le/\nqChdHPQ6PCwo3P2iP1zZT90Y9EI8RCak5odOHwW9Eh9hDqnp5gHl1ce8HvQyPL08tmtxzfA/\nB70ML5EJCQgSIQECCAkQQEiAAEICBBASIICQAAGEBAggJEAAIQECCAkQQEiAAEICBBASIICQ\nAAGEBAggJEAAIQECCAkQQEiAAEICBBASIICQAAGEBAggJEAAIQECCAkQQEiAAEICBBASIICQ\nAAGEBAggJEAAIYXNCWqV/0DR4MRh9cL2XwwyRUhhk1lIDw7poor3mr1N0+Y1/9HPuR2yPrgi\npLDJKKQ56pDLKyZ+R/3UCGnwtISlHbNAuCGksMkkpC/LDm0yHtr9SL2hhzSrQ9YFX4QUNmZI\na6fWlnQZY/6J8dfG7l7S58SPjKN/HljedfJGPaQV6hzzOdI7135gC+kEtW54+WO2sz8xsLzb\npI17DAjgmsQKIYWNEdL6PtUz5s3uVbZE05aV97j81guqun2qaUuLesy+7cQhJcY90je2pl5s\nsIZ0khp/9Oy3rWd/oajHlbefdHjFtwO6NrFBSGFjhHR68Rv6sZVVB2na7wcu1o/ONV5JOEoZ\n9zFTlf4c6VK17w2dWoY0SY3YqR9Yzn6UetX4ghrc0VcjbggpbPSQmroM/NhwpNpsfqlh2yI1\nXdtZsbfxyVtGFE3X76FU9wmLNXtIk9W9+kfL2XeW9zO+/iYhtTdCChs9pLXNr2i/q2l3H76r\ncWyatlr9wDh9WyKKxiUVexWq47+yvPz9lh7SMv0ky9lXq+GW86D9EFLY6CHVq7qnEjZqM9VB\ndy555XY9pPfVaHOgoPkHsv93tLpeD+ngMxJW6yHV6ydYzu48D9oLIYWNeY9Ul/psW0Vv4+Hd\n03pIqxL3SJtV+p0Nm4qOcTy0M0KynD15nk3cI7U3Qgob48WGLuUbjaPrNe0j9UPj2Ew9pB2l\n+xhHX9KjuKz7xsRbhKoPcwnJcvbG0lrj2F8Iqb0RUtiYr9qpC/Vj67uP0rYWHKgfe6unmqJp\nQ81X7cbrUdylppg/kJ2vpruFlD679n31nH70x4TU3ggpbIyQ1tWqk++aXVvyrKaNUlPuv6Tm\nyeJe9215sqDbBVePOqJ6sNZ4lBrwX+Xjjy3ovdYtJMvZnymonnHNyOEVhNTOCClszHc2fHx6\n7+Jdj31N/3T9+K7VRyzVftW5+8faA98s7TppY2/9Pmr79YNqVHGfM9Y6X/42Q7KcXZv/rdLd\nJ35eREjtjJDyVjb/GwUhtTdCyltzVmQ+S0jtjZBigZDaGyHFAiG1N0ICBBASIICQAAGEBAgg\nJEAAIQECCAkQQEiAAEICBBASIICQAAGEBAggJEAAIQECCAkQQEiAAEICBBASIICQAAGEBAgg\nJEAAIQECCAkQQEiAAEICBBASIICQAAH/D6SSjKlCfnhgAAAAAElFTkSuQmCC",
            "text/plain": [
              "Plot with title “Distribution: Frequencies of genes in multiple leading edges:”"
            ]
          },
          "metadata": {
            "tags": [],
            "image/png": {
              "width": 420,
              "height": 420
            },
            "text/plain": {
              "width": 420,
              "height": 420
            }
          }
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "vmfAwCsT4Mj4",
        "colab_type": "text"
      },
      "source": [
        "## Run trans eQTL analysis\n",
        "\n",
        "Let's run a trans eQTL analysis for those eight genes and the 109 AF GWAS SNPs:\n"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "vs7OGsjw2Vj_",
        "colab_type": "code",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 765
        },
        "outputId": "e856e689-d761-4d79-9441-febaf8449a69"
      },
      "source": [
        "if(file.exists(\"c5.bp.v7.1.symbols.gmt\")){\n",
        "  eQTL <- trans.qtl(prefix = \"/tmp/example\",\n",
        "                    genotype_file_name = \"1k_genomes_AF_snps.txt\",\n",
        "                    expression_file_name = t(df[, candidates]),\n",
        "                    covariates_file_name = df$sex,\n",
        "                    threshold = 1,\n",
        "                    compute.all = T,\n",
        "                    min.pv.by.genesnp = T,\n",
        "                    save.memory = F,\n",
        "                    load.qtls = T)\n",
        "  print(\"Top trans eQTLs:\")\n",
        "  head(eQTL$all$eqtls[order(eQTL$all$eqtls$pvalue), ])\n",
        "}"
      ],
      "execution_count": 11,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Created expression file from matrix for  8  genes by  272  individuals at:\n",
            " /tmp/example_gene.txt \n",
            "Created covariate file for a single gene for 272 individuals at:\n",
            " /tmp/example_covariate.txt \n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "stream",
          "text": [
            "Rows read: 109 done.\n",
            "\n",
            "Rows read: 8 done.\n",
            "\n",
            "Rows read: 1 done.\n",
            "\n",
            "Processing covariates\n",
            "\n",
            "Task finished in 0.005 seconds\n",
            "\n",
            "Processing gene expression data (imputation, residualization)\n",
            "\n",
            "Task finished in 0.011 seconds\n",
            "\n",
            "Creating output file(s)\n",
            "\n",
            "Task finished in 0.016 seconds\n",
            "\n",
            "Performing eQTL analysis\n",
            "\n",
            "100.00% done, 872 eQTLs\n",
            "\n",
            "Task finished in 0.025 seconds\n",
            "\n",
            "\n",
            "\n"
          ],
          "name": "stderr"
        },
        {
          "output_type": "stream",
          "text": [
            "Analysis of  872 QTLs done in:  0.044  seconds \n",
            "Detected eQTLs:  872 \n",
            "[1] \"Top trans eQTLs:\"\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "  snps       gene statistic pvalue       FDR       beta      \n",
              "1 rs17079881 IL10 -3.903975 0.0001196941 0.1043733 -0.2317193\n",
              "2 rs77020228 IGHD -3.591708 0.0003904660 0.1415209 -0.2915202\n",
              "3 rs332388   CD28 -3.522075 0.0005028552 0.1415209 -0.2564119\n",
              "4 rs1044258  IL10  3.450644 0.0006491785 0.1415209  0.1561586\n",
              "5 rs7526131  RGCC  3.312348 0.0010519059 0.1834524  0.1290141\n",
              "6 rs17677363 IGHD  3.224628 0.0014171063 0.2059528  0.3377210"
            ],
            "text/latex": "A data.frame: 6 × 6\n\\begin{tabular}{r|llllll}\n  & snps & gene & statistic & pvalue & FDR & beta\\\\\n  & <fct> & <fct> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n\\hline\n\t1 & rs17079881 & IL10 & -3.903975 & 0.0001196941 & 0.1043733 & -0.2317193\\\\\n\t2 & rs77020228 & IGHD & -3.591708 & 0.0003904660 & 0.1415209 & -0.2915202\\\\\n\t3 & rs332388   & CD28 & -3.522075 & 0.0005028552 & 0.1415209 & -0.2564119\\\\\n\t4 & rs1044258  & IL10 &  3.450644 & 0.0006491785 & 0.1415209 &  0.1561586\\\\\n\t5 & rs7526131  & RGCC &  3.312348 & 0.0010519059 & 0.1834524 &  0.1290141\\\\\n\t6 & rs17677363 & IGHD &  3.224628 & 0.0014171063 & 0.2059528 &  0.3377210\\\\\n\\end{tabular}\n",
            "text/markdown": "\nA data.frame: 6 × 6\n\n| <!--/--> | snps &lt;fct&gt; | gene &lt;fct&gt; | statistic &lt;dbl&gt; | pvalue &lt;dbl&gt; | FDR &lt;dbl&gt; | beta &lt;dbl&gt; |\n|---|---|---|---|---|---|---|\n| 1 | rs17079881 | IL10 | -3.903975 | 0.0001196941 | 0.1043733 | -0.2317193 |\n| 2 | rs77020228 | IGHD | -3.591708 | 0.0003904660 | 0.1415209 | -0.2915202 |\n| 3 | rs332388   | CD28 | -3.522075 | 0.0005028552 | 0.1415209 | -0.2564119 |\n| 4 | rs1044258  | IL10 |  3.450644 | 0.0006491785 | 0.1415209 |  0.1561586 |\n| 5 | rs7526131  | RGCC |  3.312348 | 0.0010519059 | 0.1834524 |  0.1290141 |\n| 6 | rs17677363 | IGHD |  3.224628 | 0.0014171063 | 0.2059528 |  0.3377210 |\n\n",
            "text/html": [
              "<table>\n",
              "<caption>A data.frame: 6 × 6</caption>\n",
              "<thead>\n",
              "\t<tr><th></th><th scope=col>snps</th><th scope=col>gene</th><th scope=col>statistic</th><th scope=col>pvalue</th><th scope=col>FDR</th><th scope=col>beta</th></tr>\n",
              "\t<tr><th></th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
              "</thead>\n",
              "<tbody>\n",
              "\t<tr><th scope=row>1</th><td>rs17079881</td><td>IL10</td><td>-3.903975</td><td>0.0001196941</td><td>0.1043733</td><td>-0.2317193</td></tr>\n",
              "\t<tr><th scope=row>2</th><td>rs77020228</td><td>IGHD</td><td>-3.591708</td><td>0.0003904660</td><td>0.1415209</td><td>-0.2915202</td></tr>\n",
              "\t<tr><th scope=row>3</th><td>rs332388  </td><td>CD28</td><td>-3.522075</td><td>0.0005028552</td><td>0.1415209</td><td>-0.2564119</td></tr>\n",
              "\t<tr><th scope=row>4</th><td>rs1044258 </td><td>IL10</td><td> 3.450644</td><td>0.0006491785</td><td>0.1415209</td><td> 0.1561586</td></tr>\n",
              "\t<tr><th scope=row>5</th><td>rs7526131 </td><td>RGCC</td><td> 3.312348</td><td>0.0010519059</td><td>0.1834524</td><td> 0.1290141</td></tr>\n",
              "\t<tr><th scope=row>6</th><td>rs17677363</td><td>IGHD</td><td> 3.224628</td><td>0.0014171063</td><td>0.2059528</td><td> 0.3377210</td></tr>\n",
              "</tbody>\n",
              "</table>\n"
            ]
          },
          "metadata": {
            "tags": []
          }
        }
      ]
    }
  ]
}